drivers/clock_control/clock_control_gd32.c

```diff
@@ -0,0 +1,298 @@
+/** RCU configuration bit (from config field) */
+#define GD32_RCU_CFG_BIT(cfg) ((cfg) & 0x1FU)
+/** RCU register (from config field) */
+#define GD32_RCU_CFG_REG(cfg) REG32(RCU + GD32_RCU_CFG_OFFSET(cfg))
+
+/** AHB prescaler exponents */
+static const uint8_t ahb_exp[] = {0U, 0U, 0U, 0U, 0U, 0U, 0U, 0U, 1U, 2U, 3U,
+				  4U, 6U, 7U, 8U, 9U};
+/** APB1 prescaler exponents */
+static const uint8_t apb1_exp[] = {0U, 0U, 0U, 0U, 1U, 2U, 3U, 4U};
+/** APB2 prescaler exponents */
+static const uint8_t apb2_exp[] = {0U, 0U, 0U, 0U, 1U, 2U, 3U, 4U};
+
+/** Reset RCU to initial state */
+static void gd32_reset_rcu_state(void)
+{
+	/* select internal 16MHz RC oscillator */
+	RCU_CTL |= RCU_CTL_IRC16MEN;
+	RCU_CFG0 = 0U;
+	while ((RCU_CFG0 & RCU_CFG0_SCSS) != RCU_SCSS_IRC16M)
+		;
+
+	/* reset HXTAL and PLL settings */
+	RCU_CTL &= ~(RCU_CTL_PLLEN | RCU_CTL_HXTALEN | RCU_CTL_CKMEN |
+		     RCU_CTL_HXTALBPS);
+	RCU_PLL = GD32_PLL_CFG_INIT;
+
+	/* disable all interrupts */
+	RCU_INT = 0U;
+}
+
+#if GD32_HXTAL_IS_ENABLED
+/** Configure and enable HXTAL */
+static void gd32_configure_hxtal(void)
+{
+	if (GD32_HXTAL_BYPASS) {
+		RCU_CTL |= RCU_CTL_HXTALBPS;
+	}
+
+	/* enable HXTAL and wait until stable */
+	RCU_CTL |= RCU_CTL_HXTALEN;
+	while ((RCU_CTL & RCU_CTL_HXTALSTB) == 0U)
+		;
+}
+#endif /* GD32_HXTAL_IS_ENABLED */
+
+#if GD32_PLL_IS_ENABLED
+/** Configure and enable PLL */
+static void gd32_configure_pll(void)
+{
+	/* configure PSC, N, P, Q and clock source */
+	RCU_PLL = GD32_PLL_CFG;
+
+	if (GD32_CK_PLL_IS_HXTAL) {
+		RCU_PLL |= RCU_PLLSRC_HXTAL;
+	} else {
+		RCU_PLL |= RCU_PLLSRC_IRC16M;
+	}
+
+	/* enable PLL and wait until stable */
+	RCU_CTL |= RCU_CTL_PLLEN;
+	while((RCU_CTL & RCU_CTL_PLLSTB) == 0U)
+		;
+}
+#endif /* GD32_PLL_IS_ENABLED */
+
+#ifdef GD32_PMU_ENABLE_HIGH_DRIVER
+/** Configure high-driver mode */
+static void gd32_enable_high_drive()
+{
+	PMU_CTL |= PMU_CTL_HDEN;
+	while((PMU_CS & PMU_CS_HDRF) == 0U)
+		;
+
+	PMU_CTL |= PMU_CTL_HDS;
+	while((PMU_CS & PMU_CS_HDSRF) == 0U)
+		;
+}
+#endif /* GD32_PMU_ENABLE_HIGH_DRIVER */
+
+/** Configure CK_SYS clock source */
+static void gd32_configure_ck_sys_source()
+{
+	RCU_CFG0 &= ~RCU_CFG0_SCS;
+
+#if GD32_CK_SYS_IS_PLL
+	RCU_CFG0 |= RCU_CKSYSSRC_PLLP;
+	while((RCU_CFG0 & RCU_SCSS_PLLP) == 0U)
+		;
+
+#if GD32_CK_PLL_IS_HXTAL
+	SystemCoreClock = GD32_HXTAL_FREQUENCY;
+#else
+	SystemCoreClock = 16000000UL;
+#endif
+	SystemCoreClock = ((SystemCoreClock / GD32_PLL_PRESCALER) *
+			   GD32_PLL_MUL_N) / GD32_PLL_DIV_P;
+#elif GD32_CK_SYS_IS_HXTAL
+	RCU_CFG0 |= RCU_CKSYSSRC_HXTAL;
+	while((RCU_CFG0 & RCU_SCSS_HXTAL) == 0U)
+		;
+
+	SystemCoreClock = GD32_HXTAL_FREQUENCY;
+#elif GD32_CK_SYS_IS_IRC16M
+	SystemCoreClock = 16000000UL;
+#endif
+
+	SystemCoreClock >>= ahb_exp[GD32_AHB_PRESCALER];
+}
+
+static int gd32_clock_control_on(const struct device *dev,
+				 clock_control_subsys_t sys)
+{
+	uint16_t cfg = *(uint16_t *)sys;
+
+	GD32_RCU_CFG_REG(cfg) |= BIT(GD32_RCU_CFG_BIT(cfg));
+
```

Comment:
I think it is a need to wait for clock stable here. 
(Or we can change this function as async_on to make compliant the API spec, but we have no method to know stabilized the clocks asynchronously... :-( )